project.clj

```diff
@@ -1,6 +1,70 @@
+  Note that these dependencies are customized somewhat when building an
+  uberjar.  See the `uberjar-dependencies` variable."
+  [['org.jruby/jruby-core jruby-version
+    :exclusions ['com.github.jnr/jffi
+                 'com.github.jnr/jnr-x86asm
+                 'org.ow2.asm/asm
+                 'org.ow2.asm/asm-analysis
+                 'org.ow2.asm/asm-commons
+                 'org.ow2.asm/asm-tree
+                 'org.ow2.asm/asm-util]]
+
+   ;; jruby-core has dependencies on discrete org.ow2.asm artifacts whereas
+   ;; other common Clojure projects like core.async declare a dependency on
+   ;; org.ow2.asm/asm-all, which provides a superset of the content of the
+   ;; discrete org.ow2.asm dependencies.  Defining asm-all here allows for
+   ;; conflict resolution to be possible in consuming projects.
+   ['org.ow2.asm/asm-all "5.0.3"]
+
+   ;; jffi and jnr-x86asm are explicit dependencies because, in JRuby's poms,
+   ;; they are defined using version ranges, and :pedantic? :abort won't
+   ;; tolerate this.
+   ['com.github.jnr/jffi jffi-version]
+   ['com.github.jnr/jffi jffi-version :classifier "native"]
+   ['com.github.jnr/jnr-x86asm "1.0.2"]
+   ['org.jruby/jruby-stdlib jruby-version]])
+
+;; For the jruby-deps uberjar builds, we want to exclude a few dependencies
+;; which we expect to be provided by other jars in the Java classpath.  We
+;; do this in order to make the dependency resolution predictable regardless
+;; of the order in which the jars are referenced on the classpath.  The
+;; following variables list the dependencies which are excluded from the
+;; default ones listed in the `base-dependencies` variable above.
+
+(def extra-top-level-dependency-exclusions-from-uberjar
+  "Set of dependencies which are excluded from the top-level ones in
+  `base-dependencies` when building an uberjar.  For example, if
+  `base-dependencies` were to include [['a] ['b] ['c] ['d]] and the set returned
+  from this variable were #{'b 'c}, the top-level dependencies built into the
+  uberjar would be [['a] ['d]]."
+  #{'org.ow2.asm/asm-all})
+
+(def extra-jruby-core-exclusions-from-uberjar
+  "Set of dependencies which are added to exclusions from jruby-core in
+  `base-dependencies` when building an uberjar.  For example, if the jruby-core
+  dependency were excluding ['a 'b] in `base-dependencies` and the set
+  returned from this variable were #{'c 'd}, the complete list of dependencies
+  which would be excluded from the jruby-core dependency would be
+  ['a 'b 'c 'd]."
+  #{'joda-time
+    'org.yaml/snakeyaml})
+
+(def uberjar-dependencies
+  "Customize the list of dependencies from `base-dependencies` for use in
+  building an uberjar."
+  (->> base-dependencies
+    (remove #(contains? extra-top-level-dependency-exclusions-from-uberjar
+               (first %)))
+    (map #(if (= 'org.jruby/jruby-core (first %))
+            (update % (dec (count %))
```

Comment:
Unrelated to the previous comments, I think it might be easier to follow the logic here if the function given to `map` was put into it's own definition so that it can have a descriptive name